Assignments/Midterm.py

```diff
@@ -0,0 +1,687 @@
+    for curr_hist, curr_group in zip(cont_pred_list, cat_resp_list):
+        fig_2.add_trace(
+            go.Violin(
+                x=np.repeat(curr_group, n),
+                y=curr_hist,
+                name=curr_group,
+                box_visible=True,
+                meanline_visible=True,
+            )
+        )
+    file = f"plots/cat_{resp}_cont_{pred}_violin_plot.html"
+    fig_2.update_layout(
+        title=f"Continuous Predictor: {pred} by Categorical Response: {resp}",
+        xaxis_title="Response",
+        yaxis_title="Predictor",
+    )
+    # fig_2.show()
+    fig_2.write_html(
+        file=f"./plots/cat_{resp}_cont_{pred}_violin_plot.html",
+        include_plotlyjs="cdn",
+    )
+    return file
+
+
+def plot_contp_contr(pred, resp, data):
+    fig = px.scatter(x=data[pred], y=data[resp], trendline="ols")
+    fig.update_layout(
+        title=f"Continuous Response: {resp} by Continuous Predictor: {pred}",
+        xaxis_title="Predictor",
+        yaxis_title="Response",
+    )
+    file = f"plots/cont_{resp}_cont_{pred}_scatter_plot.html"
+    # fig.show()
+    fig.write_html(
+        file=f"./plots/cont_{resp}_cont_{pred}_scatter_plot.html",
+        include_plotlyjs="cdn",
+    )
+    return file
+
+
+def mean_diff_tbl(pred, resp, data):
+    df = pd.DataFrame(
+        columns=[
+            "(i)",
+            "Bin_Pop",
+            "Bin_Mean",
+            "Pop_Mean",
+            "Pop_Proportion",
+            "Mean_Sq_Diff",
+            "Mean_Sq_Diff_Weighted",
+        ]
+    )
+    pop_mean = data[resp].mean()
+    if not is_continuous(data, pred):
+        for i, val in enumerate(data[pred].unique()):
+            bin_mean = data[resp][data[pred] == val].mean()
+            bin_pop = len(data[resp][data[pred] == val])
+            pop_prop = bin_pop / len(data[resp])
+            mean_sq_diff = (pop_mean - bin_mean) ** 2
+            mean_sq_diff_w = mean_sq_diff * pop_prop
+            df.loc[i] = [
+                val,
+                bin_pop,
+                bin_mean,
+                pop_mean,
+                pop_prop,
+                mean_sq_diff,
+                mean_sq_diff_w,
+            ]
+    else:
+        data_sorted = data.sort_values(pred)
+        data_sorted["bins"] = pd.cut(data_sorted[pred], 10)
+        for i, cur_bin in enumerate(data_sorted["bins"].unique()):
+            bin_mean = data_sorted[resp][data_sorted["bins"] == cur_bin].mean()
+            bin_pop = len(data_sorted[pred][data_sorted["bins"] == cur_bin])
+            pop_prop = bin_pop / len(data_sorted[pred])
+            mean_sq_diff = (pop_mean - bin_mean) ** 2
+            mean_sq_diff_w = mean_sq_diff * pop_prop
+            df.loc[i] = [
+                i,
+                bin_pop,
+                bin_mean,
+                pop_mean,
+                pop_prop,
+                mean_sq_diff,
+                mean_sq_diff_w,
+            ]
+
+    return df
+
+
+def brute_force(pred1, pred2, resp, data):
```

Comment:
I think this would be easier if you kept the data in pandas. I see what you're doing with the np.zeros. Also, you could probably simplify this by not doing 2 nested ifs. I think you just need one if to determine the type of each predictor. You've got them nested. Maybe just set the bins based on the predictor type, then move to the next one and keep the bulk of the logic outside those 2 ifs. (meaning after those 2 ifs).